algorithms/src/snark/marlin/ahp/prover/round_functions/first.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use core::convert::TryInto;
+
+use crate::{
+    fft::{DensePolynomial, EvaluationDomain, Evaluations as EvaluationsOnDomain, SparsePolynomial},
+    polycommit::sonic_pc::{LabeledPolynomial, LabeledPolynomialWithBasis, PolynomialWithBasis},
+    snark::marlin::{
+        ahp::{indexer::CircuitInfo, AHPError, AHPForR1CS},
+        prover,
+        MarlinMode,
+    },
+};
+use snarkvm_fields::PrimeField;
+use snarkvm_utilities::cfg_into_iter;
+
+use rand_core::RngCore;
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+    /// Output the number of oracles sent by the prover in the first round.
+    pub fn prover_num_first_round_oracles() -> usize {
+        if MM::ZK { 4 } else { 3 }
+    }
+
+    /// Output the degree bounds of oracles in the first round.
+    pub fn prover_first_round_degree_bounds(_info: &CircuitInfo<F>) -> impl Iterator<Item = Option<usize>> {
+        if MM::ZK { core::iter::repeat(None).take(4) } else { core::iter::repeat(None).take(3) }
+    }
+
+    /// Output the first round message and the next state.
+    #[allow(clippy::type_complexity)]
+    pub fn prover_first_round<'a, R: RngCore>(
+        mut state: prover::State<'a, F, MM>,
+        rng: &mut R,
+    ) -> Result<(prover::FirstOracles<'a, F>, prover::State<'a, F, MM>), AHPError> {
+        let round_time = start_timer!(|| "AHP::Prover::FirstRound");
+        let constraint_domain = state.constraint_domain;
+        let zk_bound = state.zk_bound;
+
+        let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(3);
+        let z_a = state.z_a.take().unwrap();
+        let z_b = state.z_b.take().unwrap();
+        let mut private_variables = vec![];
+        core::mem::swap(&mut state.private_variables, &mut private_variables);
```

Comment:
nit: `mem::take` does the same thing (as long as there are no `Default` impl issues)